data/input/parameters.txt
### Mathematical model's parameters ###

# Domain [x_min, x_max] \times [y_min, y_max] \times [z_min, z_max] \times [t_min, t_max] #
x_min = -500
x_max = 700
y_min = -200
y_max = 400
z_min = 0
z_max = 5
t_min = 0
t_max = 30

# Numerical grid #
Nx = 512
Ny = 256
Nz = 256
Nt = 50000
NT = 5000

# Time numerical method
method = RK2

# Atmosphere properties #
# Specific heat capacity (constant pressure) in J kg^{-1} K^{-1} or m^2 s^{-2} K^{-1} (Air: 1007 at 15 °C, NASA - Cengel 2018)
c_p = 1007 
# Specific heat capacity (constant volume) in J kg^{-1} K^{-1} or m^2 s^{-2} K^{-1} (Air: 717)
c_V =  717 
# Density in kg m^{-3} (Air: 1.225 at 15 °C - Cengel 2018)
rho = 1.225 
# Dynamic viscosity in kg m^{-1} s^{-1} (Air: 1.802e-5, Cengel 2018)
mu = 1.802e-5 

# Temperaure equation parameters #
# Thermal conductivity in W m^{-1} K^{-1} or kg m s^{-3} K ^{-1}  (Air: 0.02476 at 15 °C, NASA - Cengel 2018)
k = 0.02476 
# Thermal diffusivity in m^2 s^{-1} (Air: 2.009e-5 at 15 °C - NASA)
alpha = 2.009e-5
# Optical path length in m
delta = 1 

# Fluid equations parameters #
# Kinematic viscosity in m^2 s^{-1} (Air: 1.47e-5, Cengel 2018)
nu = 1.47e-5
# Temperature of the environment in K (Ambient temperature: 15°C - 288.15 K, NASA - Cengel 2018)
T_inf = 288.15 
# Acceleration due to gravity in m s^{-2} (Typical value: 9.81 m s^{-2})
g = -9.81 
# Turbulence model
turb = True 
# Conservative equation for convection
conser = False 
# Smagorinsky constant (McGrattan 2023)
C_s = 0.2 
# Prandtl number 1. / (Air: ~.74)
Pr = 0.732
# Drag coefficient "1 or near to unity according to works of Mell and Linn" 1
C_D = 1 
#  Contact area per unit volume between the gas and the solid in m^{-1} (6000 in Mell or Linn)
a_v = 5.508 

# Fuel and reaction parameters #
# Temperature of solid-gas phase change in K. (473 - 573 K)
T_pc = 523
# Heat energy per unit of mass (wood) in J kg^{-1} or m^2 s^{-2}. 
# About 21.20e6 for wood according to https://en.wikipedia.org/wiki/Heat_of_combustion
# 15.6 (Mell 2007 - FDS), 19.4 (Dupuy 2011 - FIRETEC)
H_R = 21.2e6 
# Pre-exponential factor in s^{-1}. (1e9, Asensio 2002)
A = 1e9 
# Activation energy in J mol^{-1} or kg m^2 s^{-2} mol^{-1}. E_A = 20e3 cal mol^{-1} according to (Asensio 2002). 
E_A = 150e3 
# Activation temperature in K 
T_a = 18040.8533 
# Convection coefficient in W m^{-2} K^{-1} or kg s^{-3} K^{-1}  (Air: 0.5-1000), (15.9 - 18.2, Maragkos 2021) 2D: 1.147
h = 1.147
# Threshold to add solid fuel force 0.04 in 2D experiment
Y_D = 0.04
# Extra parameter to control the fuel consumption rate 100 in 2D experiment
Y_f = 100
# Temperature of initial fire source in K 788.15
T_hot = 650
# Height of fuel in m
Y_h = 0.51

#include_source = True
#source_filter = False
#radiation = False
#sutherland_law = False
#debug_pde = False
#bound = True
# Temperature bounds
T_min = 288.15 
T_max = 1000
# Fuel fraction bounds
Y_min = 0 
Y_max = 1
# Temperature source for t time
# t_source = -1 # If t_source < 0, then the temperature source is not used

### Initial conditions parameters ###
# Wind# 'power law' or 'log'
U0_type = power_law
# Log wind profile
# Surface roughness in m
u_z0 = 0.05 
# Zero-plane displacement in m
d = 0 
# Friction velocity in ms^{-1}
u_ast = .1
# Von Karman constant in 1
k = 0.41 
# Power law (used by FDS)
# Reference speed in ms^{-1}
u_r = 4.8 
# Reference height in m
z_r = 2 
# Empirical constant in 1
alpha_u = 0.14285714285

# Temperature 
# Shape 'plate' or 'gaussian'
T0_shape = gaussian
# Location
# 'Width' of initial fire source in m
T0_x_start = 0
# PLATE -> x_start + 3.3 (FDS), 6 example
T0_x_end = 6
T0_y_start = 80
T0_y_end = 120 
# 'Height' of initial fire source in m
T0_z_start = 0
T0_z_end = 1 

# Topography
# 'flat' or 'hill'
Z_shape = 'flat'
# Center of hill in m
hill_center_x = 50 
# Center of hill in m
hill_center_y = 50
# Length of hill in m 
hill_length = 20 
# Width of hill in m
hill_width = 20 
# Height of hill in m
hill_height = 2.5 

### Immerse boundary method parameterns ###
u_dead_nodes = 0
v_dead_nodes = 0
w_dead_nodes = 0
T_dead_nodes = 288.15
Y_dead_nodes = 0

dead_nodes_values = [
    u_dead_nodes, 
    v_dead_nodes, 
    w_dead_nodes, 
    T_dead_nodes, 
    Y_dead_nodes
]

# Sutherland's law parameters
S_T_0 = 273 # Reference temperature in K
S_k_0 = 0.024 # Thermal conductivity in W m^{-1} K^{-1} or kg m s^{-3} K ^{-1}  (Air: 0.024)
S_k = 194 # Sutherland's constant in K

spatial_dims = 2

periodic_axes = (False, True)

output_dir = './data/output/'


src/c/parameters.c
/**
 * @file parameters.c
 * @author Daniel San Martin (dsanmartinreyes@gmail.com)
 * @brief Functions to handle the simulation parameters from a file.
 * @version 0.1
 * @date 2024-07-21
 * 
 * @copyright Copyright (c) 2024
 * 
 */
#include "../../include/c/parameters.h"

Parameters read_parameters_file(const char *file_path) {
    FILE *parameters_file = fopen(file_path, "r"); // Open file
    int size; // Size of the fields
    char line[MAX_LINE_LENGTH]; // Create line buffer
    char parameter_file_path[256]; // Path to the parameters file
    char log_file_path[256]; // Path to the log file
    Parameters parameters; // Create parameters struct
    memset(&parameters, 0, sizeof(Parameters)); // Initialize parameters struct
    generate_current_datetime_string(parameters.sim_id, 32); // Get YYYYMMDD for simulation ID
    strcpy(parameters.save_path, "data/output/"); // Set save path
    // Concatenate simulation ID to save path
    strcat(parameters.save_path, parameters.sim_id);
    strcat(parameters.save_path, "/");
    // Create directory if it does not exist
    mkdir(parameters.save_path, 0777);
    // Concatenate save path to parameters file path
    strcpy(parameter_file_path, parameters.save_path);
    strcat(parameter_file_path, "parameters.txt");
    // Create file for parameters.txt
    parameters.log_files.parameters = fopen(parameter_file_path, "w");
    // Concatenate save path to log file path
    strcpy(log_file_path, parameters.save_path);
    strcat(log_file_path, "log.txt");
    // Create file for log.txt
    parameters.log_files.log = fopen(log_file_path, "w");
    // Read file line by line
    while (fgets(line, MAX_LINE_LENGTH, parameters_file) != NULL) {
        if (strncmp(line, "Nx =", 4) == 0) {
            sscanf(line + 5, "%d", &(parameters.Nx));
        } 
        if (strncmp(line, "Ny =", 4) == 0) {
            sscanf(line + 5, "%d", &(parameters.Ny));
        }
        if (strncmp(line, "Nz =", 4) == 0) {
            sscanf(line + 5, "%d", &(parameters.Nz));
        }
        if (strncmp(line, "Nt =", 4) == 0) {
            sscanf(line + 5, "%d", &(parameters.Nt));
        }
        if (strncmp(line, "NT =", 4) == 0) {
            sscanf(line + 5, "%d", &(parameters.NT));
        }
        if (strncmp(line, "x_min =", 7) == 0) {
            sscanf(line + 8, "%lf", &(parameters.x_min));
        }
        if (strncmp(line, "x_max =", 7) == 0) {
            sscanf(line + 8, "%lf", &(parameters.x_max));
        }
        if (strncmp(line, "y_min =", 7) == 0) {
            sscanf(line + 8, "%lf", &(parameters.y_min));
        }
        if (strncmp(line, "y_max =", 7) == 0) {
            sscanf(line + 8, "%lf", &(parameters.y_max));
        }
        if (strncmp(line, "z_min =", 7) == 0) {
            sscanf(line + 8, "%lf", &(parameters.z_min));
        }
        if (strncmp(line, "z_max =", 7) == 0) {
            sscanf(line + 8, "%lf", &(parameters.z_max));
        }
        if (strncmp(line, "t_min =", 7) == 0) {
            sscanf(line + 8, "%lf", &(parameters.t_min));
        }
        if (strncmp(line, "t_max =", 7) == 0) {
            sscanf(line + 8, "%lf", &(parameters.t_max));
        } 
        if (strncmp(line, "nu =", 4) == 0) {
            sscanf(line + 5, "%lf", &(parameters.nu));
        } 
        if (strncmp(line, "u_r =", 5) == 0) {
            sscanf(line + 6, "%lf", &(parameters.u_r));
        }
        if (strncmp(line, "z_r =", 5) == 0) {
            sscanf(line + 6, "%lf", &(parameters.z_r));
        }
        if (strncmp(line, "alpha_u =", 9) == 0) {
            sscanf(line + 10, "%lf", &(parameters.alpha_u));
        }
        if (strncmp(line, "alpha =", 7) == 0) {
            sscanf(line + 8, "%lf", &(parameters.alpha));
        }
        if (strncmp(line, "T_inf =", 7) == 0) {
            sscanf(line + 8, "%lf", &(parameters.T_inf));
        }
        if (strncmp(line, "T_hot =", 7) == 0) {
            sscanf(line + 8, "%lf", &(parameters.T_hot));
        }
        if (strncmp(line, "T0_shape =", 10) == 0) {
            sscanf(line + 11, "%s", parameters.T0_shape);
        }
        if (strncmp(line, "T0_x_start =", 12) == 0) {
            sscanf(line + 13, "%lf", &(parameters.T0_x_start));
        }
        if (strncmp(line, "T0_x_end =", 10) == 0) {
            sscanf(line + 11, "%lf", &(parameters.T0_x_end));
        }
        // if (strncmp(line, "T0_x_center =", 13) == 0) {
        //     sscanf(line + 14, "%lf", &(parameters.T0_x_center));
        // }
        // if (strncmp(line, "T0_length =", 11) == 0) {
        //     sscanf(line + 12, "%lf", &(parameters.T0_length));
        // }
        if (strncmp(line, "T0_y_start =", 12) == 0) {
            sscanf(line + 13, "%lf", &(parameters.T0_y_start));
        }
        if (strncmp(line, "T0_y_end =", 10) == 0) {
            sscanf(line + 11, "%lf", &(parameters.T0_y_end));
        }
        // if (strncmp(line, "T0_y_center =", 13) == 0) {
        //     sscanf(line + 14, "%lf", &(parameters.T0_y_center));
        // }
        // if (strncmp(line, "T0_width =", 10) == 0) {
        //     sscanf(line + 11, "%lf", &(parameters.T0_width));
        // }
        if (strncmp(line, "T0_z_start =", 12) == 0) {
            sscanf(line + 13, "%lf", &(parameters.T0_z_start));
        }
        if (strncmp(line, "T0_z_end =", 10) == 0) {
            sscanf(line + 11, "%lf", &(parameters.T0_z_end));
        }
        // if (strncmp(line, "T0_z_center =", 13) == 0) {
        //     sscanf(line + 14, "%lf", &(parameters.T0_z_center));
        // }
        // if (strncmp(line, "T0_height =", 11) == 0) {
        //     sscanf(line + 12, "%lf", &(parameters.T0_height));
        // }
        if (strncmp(line, "Y_h =", 5) == 0) {
            sscanf(line + 6, "%lf", &(parameters.Y_h));
        }
        if (strncmp(line, "Y_f =", 5) == 0) {
            sscanf(line + 6, "%lf", &(parameters.Y_f));
        }
        // Fuel drag Y_D
        if (strncmp(line, "Y_D =", 5) == 0) {
            sscanf(line + 6, "%lf", &(parameters.Y_D));
        }
        // Read H_R 
        if (strncmp(line, "H_R =", 5) == 0) {
            sscanf(line + 6, "%lf", &(parameters.H_R));
        }
        // Read A
        if (strncmp(line, "A =", 3) == 0) {
            sscanf(line + 4, "%lf", &(parameters.A));
        }
        // Read activation temperature T_a
        if (strncmp(line, "T_a =", 5) == 0) {
            sscanf(line + 6, "%lf", &(parameters.T_a));
        }
        // Temperature phase-change T_pc
        if (strncmp(line, "T_pc =", 6) == 0) {
            sscanf(line + 7, "%lf", &(parameters.T_pc));
        }
        // Convective heat transfer coefficient h
        if (strncmp(line, "h =", 3) == 0) {
            sscanf(line + 4, "%lf", &(parameters.h));
        }
        // Volumetric heat capacity a_v
        if (strncmp(line, "a_v =", 5) == 0) {
            sscanf(line + 6, "%lf", &(parameters.a_v));
        }
        // Heat capacity c_p
        if (strncmp(line, "c_p =", 5) == 0) {
            sscanf(line + 6, "%lf", &(parameters.c_p));
        }
        // Density rho
        if (strncmp(line, "rho =", 5) == 0) {
            sscanf(line + 6, "%lf", &(parameters.rho));
        }
        // U0 Type
        if (strncmp(line, "U0_type =", 9) == 0) {
            sscanf(line + 10, "%s", (char *) &(parameters.U0_type));
        }
        // T0 Shape
        if (strncmp(line, "T0_shape =", 10) == 0) {
            sscanf(line + 11, "%s", (char *) &(parameters.T0_shape));
        }
        // Acceleration due to gravity g
        if (strncmp(line, "g =", 3) == 0) {
            sscanf(line + 4, "%lf", &(parameters.g));
        }
        // Smagorinsky constant C_s
        if (strncmp(line, "C_s =", 5) == 0) {
            sscanf(line + 6, "%lf", &(parameters.C_s));
        }
        // Prandtl number Pr
        if (strncmp(line, "Pr =", 4) == 0) {
            sscanf(line + 5, "%lf", &(parameters.Pr));
        }
        // p_top
        if (strncmp(line, "p_top =", 7) == 0) {
            sscanf(line + 8, "%lf", &(parameters.p_top));
        }
        // Time integration method
        if (strncmp(line, "method =", 8) == 0) {
            sscanf(line + 9, "%s", (char *) &(parameters.method));
        }
        // Number of threads
        if (strncmp(line, "threads =", 9) == 0) {
            sscanf(line + 10, "%d", &(parameters.n_threads));
        }
        // Temperature bounds
        if (strncmp(line, "T_min =", 6) == 0) {
            sscanf(line + 7, "%lf", &(parameters.T_min));
        }
        if (strncmp(line, "T_max =", 6) == 0) {
            sscanf(line + 7, "%lf", &(parameters.T_max));
        }
        // Fuel bounds
        if (strncmp(line, "Y_min =", 6) == 0) {
            sscanf(line + 7, "%lf", &(parameters.Y_min));
        }
        if (strncmp(line, "Y_max =", 6) == 0) {
            sscanf(line + 7, "%lf", &(parameters.Y_max));
        }
        // Topography shape
        if (strncmp(line, "topo_shape =", 11) == 0) {
            sscanf(line + 12, "%s", (char *) &(parameters.topo_shape));
        }
        // Dead nodes values
        if (strncmp(line, "u_dead_nodes =", 13) == 0) {
            sscanf(line + 14, "%lf", &(parameters.u_dead_nodes));
        }
        if (strncmp(line, "v_dead_nodes =", 13) == 0) {
            sscanf(line + 14, "%lf", &(parameters.v_dead_nodes));
        }
        if (strncmp(line, "w_dead_nodes =", 13) == 0) {
            sscanf(line + 14, "%lf", &(parameters.w_dead_nodes));
        }
        if (strncmp(line, "T_dead_nodes =", 13) == 0) {
            sscanf(line + 14, "%lf", &(parameters.T_dead_nodes));
        }
        if (strncmp(line, "Y_dead_nodes =", 13) == 0) {
            sscanf(line + 14, "%lf", &(parameters.Y_dead_nodes));
        }
        // Hill parameters
        if (strncmp(line, "hill_center_x =", 14) == 0) {
            sscanf(line + 15, "%lf", &(parameters.hill_center_x));
        }
        if (strncmp(line, "hill_center_y =", 14) == 0) {
            sscanf(line + 15, "%lf", &(parameters.hill_center_y));
        }
        if (strncmp(line, "hill_length =", 12) == 0) {
            sscanf(line + 13, "%lf", &(parameters.hill_length));
        }
        if (strncmp(line, "hill_width =", 11) == 0) {
            sscanf(line + 12, "%lf", &(parameters.hill_width));
        }
        if (strncmp(line, "hill_height =", 12) == 0) {
            sscanf(line + 13, "%lf", &(parameters.hill_height));
        }
    }
    parameters.alpha_u = 1.0 / 7.0;
    // Initialize x, y, z, t
    parameters.x = (double *) malloc(parameters.Nx * sizeof(double));
    parameters.y = (double *) malloc(parameters.Ny * sizeof(double));
    parameters.z = (double *) malloc(parameters.Nz * sizeof(double));
    parameters.t = (double *) malloc((parameters.Nt + 1) * sizeof(double));
    // Initialize r, s
    parameters.r = (double *) malloc((parameters.Nx - 1) * sizeof(double));
    parameters.s = (double *) malloc((parameters.Ny - 1) * sizeof(double));
    parameters.kx = (double *) malloc((parameters.Nx - 1) * sizeof(double));
    parameters.ky = (double *) malloc((parameters.Ny - 1) * sizeof(double));
    // IBM malloc
    // Allocate memory
    parameters.cut_nodes = (int *)malloc(parameters.Nx * parameters.Ny * sizeof(int)); // Cut nodes
    parameters.z_ibm = (double *)malloc(parameters.Nx * parameters.Ny * parameters.Nz * sizeof(double)); // IBM z distance
    parameters.topography = (double *)malloc(parameters.Nx * parameters.Ny * sizeof(double)); // Topography
    // Nz for Y variable. To store the number of grid points in z for Y variable
    parameters.Nz_Y = (int *)malloc(parameters.Nx * parameters.Ny * sizeof(int));
    // Compute dx, dy, dz, dts
    parameters.dx = (parameters.x_max - parameters.x_min) / (parameters.Nx - 1);
    parameters.dy = (parameters.y_max - parameters.y_min) / (parameters.Ny - 1);
    parameters.dz = (parameters.z_max - parameters.z_min) / (parameters.Nz - 1);
    parameters.dt = (parameters.t_max - parameters.t_min) / parameters.Nt;
    // parameters.Nz_Y_max = 0;
    // Frequency domain
    fft_freq(parameters.r, parameters.Nx - 1, 1.0 / (parameters.Nx - 1));
    fft_freq(parameters.s, parameters.Ny - 1, 1.0 / (parameters.Ny - 1));
    // Physical domain and frequency domain
    for (int i = 0; i < parameters.Nx; i++) {
        parameters.x[i] = parameters.x_min + i * parameters.dx;
        if (i < parameters.Nx - 1) {
            parameters.kx[i] = 2 * M_PI * parameters.r[i] * parameters.dz / (parameters.x_max - parameters.x_min);
        }
    }
    for (int j = 0; j < parameters.Ny; j++) {
        parameters.y[j] = parameters.y_min + j * parameters.dy;
        if (j < parameters.Ny - 1) {
            parameters.ky[j] = 2 * M_PI * parameters.s[j] * parameters.dz / (parameters.y_max - parameters.y_min);
        }
    }
    for (int k = 0; k < parameters.Nz; k++) {
        parameters.z[k] = parameters.z_min + k * parameters.dz;
    }
    for (int n = 0; n <= parameters.Nt; n++) {
        parameters.t[n] = parameters.t_min + n * parameters.dt;
    }
    // Computer T0 centers and dimensiones
    parameters.T0_x_center = (parameters.T0_x_start + parameters.T0_x_end) / 2;
    parameters.T0_y_center = (parameters.T0_y_start + parameters.T0_y_end) / 2;
    parameters.T0_z_center = 0.0;
    parameters.T0_length = parameters.T0_x_end - parameters.T0_x_start;
    parameters.T0_width = parameters.T0_y_end - parameters.T0_y_start;
    parameters.T0_height = parameters.T0_z_end - parameters.T0_z_start;
    // IBM topography
    if (strcmp(parameters.topo_shape, "hill") == 0)
        simple_hill(&parameters);
    else if (strcmp(parameters.topo_shape, "flat") == 0)
        flat_terrain(&parameters);
    ibm_parameters(&parameters);    
    // Sizes
    size = parameters.Nx * parameters.Ny * parameters.Nz;
    // Fill field indexes
    parameters.field_indexes.u = 0;
    parameters.field_indexes.v = size;
    parameters.field_indexes.w = 2 * size;
    parameters.field_indexes.T = 3 * size;
    parameters.field_indexes.Y = 4 * size;
    // Fill turbulence indexes
    parameters.turbulence_indexes.ux = 0;
    parameters.turbulence_indexes.uy = size;
    parameters.turbulence_indexes.uz = 2 * size;
    parameters.turbulence_indexes.vx = 3 * size;
    parameters.turbulence_indexes.vy = 4 * size;
    parameters.turbulence_indexes.vz = 5 * size;
    parameters.turbulence_indexes.wx = 6 * size;
    parameters.turbulence_indexes.wy = 7 * size;
    parameters.turbulence_indexes.wz = 8 * size;
    parameters.turbulence_indexes.Tx = 9 * size;
    parameters.turbulence_indexes.Ty = 10 * size;
    parameters.turbulence_indexes.Tz = 11 * size;
    parameters.turbulence_indexes.uxx = 12 * size;
    parameters.turbulence_indexes.uyy = 13 * size;
    parameters.turbulence_indexes.uzz = 14 * size;
    parameters.turbulence_indexes.vxx = 15 * size;
    parameters.turbulence_indexes.vyy = 16 * size;
    parameters.turbulence_indexes.vzz = 17 * size;
    parameters.turbulence_indexes.wxx = 18 * size;
    parameters.turbulence_indexes.wyy = 19 * size;
    parameters.turbulence_indexes.wzz = 20 * size;
    parameters.turbulence_indexes.Txx = 21 * size;
    parameters.turbulence_indexes.Tyy = 22 * size;
    parameters.turbulence_indexes.Tzz = 23 * size;
    parameters.turbulence_indexes.fw = 24 * size;
    // Close file
    fclose(parameters_file);
    return parameters;
}

void free_parameters(Parameters *parameters) {
    free(parameters->x);
    free(parameters->y);
    free(parameters->z);
    free(parameters->t);
    free(parameters->r);
    free(parameters->s);
    free(parameters->kx);
    free(parameters->ky);
    free(parameters->cut_nodes);
    free(parameters->z_ibm);
    free(parameters->topography);
    free(parameters->Nz_Y);
    fclose(parameters->log_files.parameters);
    fclose(parameters->log_files.log);
}
